crates/solvers/src/domain/solver/dex.rs

```diff
@@ -0,0 +1,77 @@
+//! A simple solver that matches orders directly with swaps from the external
+//! DEX and DEX aggregator APIs.
+
+use {
+    crate::{
+        domain::{
+            auction,
+            dex::{self, slippage},
+            order, solution,
+        },
+        infra,
+    },
+    tracing::Instrument,
+};
+
+pub struct Dex {
+    /// The DEX API client.
+    pub dex: infra::dex::Dex,
+
+    /// The slippage configuration to use for the solver.
+    pub slippage: slippage::Limits,
+}
+
+impl Dex {
+    pub async fn solve(&self, auction: auction::Auction) -> Vec<solution::Solution> {
+        // TODO: order prioritization, skip liquidity orders, concurrency.
+        let prices = slippage::Prices::for_auction(&auction);
+
+        let mut solutions = Vec::new();
+        for order in auction.orders {
+            let span = tracing::info_span!("solve", order = %order.uid);
+            if let Some(solution) = self
+                .solve_order(order, &prices, auction.gas_price)
+                .instrument(span)
+                .await
+            {
+                solutions.push(solution);
+            }
+        }
+
+        solutions
+    }
+
+    async fn solve_order(
+        &self,
+        order: order::Order,
+        prices: &slippage::Prices,
+        gas: auction::GasPrice,
+    ) -> Option<solution::Solution> {
+        let query = dex::Order::new(&order);
+        let slippage = self.slippage.relative(&query.amount(), prices);
+
+        let swap = match self.dex.swap(&query, &slippage, gas).await {
+            Ok(swap) => swap,
+            Err(infra::dex::Error::NotFound) => {
+                tracing::debug!("cound not find swap for order");
+                return None;
+            }
+            Err(infra::dex::Error::OrderNotSupported) => {
+                tracing::debug!("skipping unsupported order");
+                return None;
+            }
+            Err(infra::dex::Error::Other(err)) => {
+                tracing::warn!(?err, "failed to get swap");
+                return None;
+            }
+        };
+
+        let Some(solution) = swap.into_solution(order) else {
+            tracing::debug!("no solution for swap");
+            return None;
+        };
+
+        tracing::debug!("solved");
+        Some(solution)
```

Comment:
We talked a bit [in Slack](https://cowservices.slack.com/archives/C049P351GKF/p1677062705741459) about trying to avoid logs in the domain logic.

Notably, logs for the different `Dex::swap` results could in theory be in `infra`. However, I was a bit torn about putting it there as I explicitly want to log every outcome of solving an order (for observability and debugging purposes so we can trace an order from E2E in the logs). With that in mind, it felt a bit weird to log some of the outcomes in `infra` (and rely on logging implementation details of that module), and the outcome of `Swap::into_solution` here (which is domain-logic and can't be pushed into `infra`). Hence, I decided to put the logging of the outcomes of `solve_order` into the domain.

Reply:
That is the plan.